power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -0,0 +1,181 @@
+                                     total_elements, indptr, data);
+        handle.check_error();
+    }
+    void add_buffer(std::string const& component, Idx elements_per_scenario, Idx total_elements, Idx const* indptr,
+                    void const* data) {
+        return add_buffer(handle_, *this, component, elements_per_scenario, total_elements, indptr, data);
+    }
+
+    static void get_info(Handle const& handle, DatasetConst const& dataset) {
+        PGM_dataset_const_get_info(handle.get(), dataset.get());
+        handle.check_error();
+    }
+    void get_info() const { return get_info(handle_, *this); }
+
+  private:
+    UniquePtr<PGM_ConstDataset, PGM_destroy_dataset_const> dataset_;
+};
+
+class DatasetWritable : public Dataset {
+  public:
+    DatasetWritable() : Dataset() {}
+
+    static PGM_DatasetInfo const* get_info(Handle const& handle, PGM_WritableDataset const* dataset) {
+        return PGM_dataset_writable_get_info(handle.get(), dataset);
+        handle.check_error();
+    }
+    PGM_DatasetInfo const* get_info(PGM_WritableDataset const* dataset) const { return get_info(handle_, dataset); }
+
+    static void set_buffer(Handle const& handle, PGM_WritableDataset* dataset, std::string const& component,
+                           Idx* indptr, void* data) {
+        PGM_dataset_writable_set_buffer(handle.get(), dataset, component.c_str(), indptr, data);
+        handle.check_error();
+    }
+    void set_buffer(PGM_WritableDataset* dataset, std::string const& component, Idx* indptr, void* data) {
+        return set_buffer(handle_, dataset, component.c_str(), indptr, data);
+    }
+};
+
+class DatasetMutable : public Dataset {
+    DatasetMutable(std::string const& dataset, Idx is_batch, Idx batch_size)
+        : Dataset(), dataset_{PGM_create_dataset_mutable(handle_.get(), dataset.c_str(), is_batch, batch_size)} {}
+
+    PGM_MutableDataset* get() const { return dataset_.get(); }
+
+    static void add_buffer(Handle const& handle, DatasetMutable& dataset, std::string const& component,
+                           Idx elements_per_scenario, Idx total_elements, Idx const* indptr, void* data) {
+        PGM_dataset_mutable_add_buffer(handle.get(), dataset.get(), component.c_str(), elements_per_scenario,
+                                       total_elements, indptr, data);
+        handle.check_error();
+    }
+    void add_buffer(std::string const& component, Idx elements_per_scenario, Idx total_elements, Idx const* indptr,
+                    void* data) {
+        return add_buffer(handle_, *this, component.c_str(), elements_per_scenario, total_elements, indptr, data);
+    }
+
+    static PGM_DatasetInfo const* get_info(Handle const& handle, DatasetMutable const& dataset) {
+        PGM_dataset_mutable_get_info(handle.get(), dataset.get());
+        handle.check_error();
+    }
+    PGM_DatasetInfo const* get_info() const { return get_info(handle_, *this); }
```

Comment:
i don't think we should explicitly return a raw C API pointer here (or anywhere, for all that matters). even if it's just an alias, that would make things better. e.g.:

```cpp
namespace power_grid_model_cpp {
using DatasetInfo = PGM_DatasetInfo;
} // namespace power_grid_model_cpp
```

Reply:
You are right. Will implement this.